py4DSTEM/process/phase/parameter_optimize.py

```diff
@@ -0,0 +1,449 @@
+        non-optimization arguments and accepts a concatenated set of keyword arguments. The
+        wrapper function returns the final error value from the ptychography run.
+
+        parameter_list is a list of skopt Dimension objects
+
+        Both static and optimization args are passed in dictionaries. The values of the
+        static dictionary are the fixed parameters, and only the keys of the optimization
+        dictionary are used.
+        """
+
+        # Get lists of optimization parameters for each step
+        init_params = list(init_optimization_params.keys())
+        afft_params = list(affine_optimization_params.keys())
+        prep_params = list(preprocess_optimization_params.keys())
+        reco_params = list(reconstruct_optimization_params.keys())
+
+        # Construct partial methods to encapsulate the static parameters.
+        # If only ``reconstruct`` has optimization variables, perform
+        # preprocessing now, store the ptycho object, and use dummy
+        # functions instead of the partials
+        if (len(init_params), len(afft_params), len(prep_params)) == (0, 0, 0):
+            affine_preprocessed = AffineTransform(**affine_static_args)
+            init_args = init_static_args.copy()
+            init_args["initial_scan_positions"] = self._get_scan_positions(
+                affine_preprocessed, init_static_args["datacube"]
+            )
+
+            ptycho_preprocessed = cls(**init_args).preprocess(
+                **preprocess_static_args
+            )
+
+            def obj(**kwargs):
+                return ptycho_preprocessed
+
+            def prep(ptycho, **kwargs):
+                return ptycho
+
+        else:
+            obj = partial(cls, **init_static_args)
+            prep = partial(cls.preprocess, **preprocess_static_args)
+
+        affine = partial(AffineTransform, **affine_static_args)
+        recon = partial(cls.reconstruct, **reconstruct_static_args)
+        
+        # Target function for Gaussian process optimization that takes a single
+        # dict of named parameters and returns the ptycho error metric
+        @use_named_args(parameter_list)
+        def f(**kwargs):
+            init_args = {k: kwargs[k] for k in init_params}
+            afft_args = {k: kwargs[k] for k in afft_params}
+            prep_args = {k: kwargs[k] for k in prep_params}
+            reco_args = {k: kwargs[k] for k in reco_params}
+
+            # Create affine transform object
+            tr = affine(**afft_args)
+            # Apply affine transform to pixel grid, using the
+            # calibrations lifted from the dataset
+            dataset = init_static_args["datacube"]
+            init_args["initial_scan_positions"] = self._get_scan_positions(tr, dataset)
+
+            ptycho = obj(**init_args)
+            prep(ptycho, **prep_args)
+            recon(ptycho, **reco_args)
+
+            return np.log(ptycho.error)
```

Comment:
Error divergence check and log error suggestion:

```suggestion
            errors = ptycho.error_iterations
            if check_error_divergence and errors[-1] > min(errors[:-1]):
                error = 1.0
            else:
                error = errors[-1]
            
            if return_log_error:
                error = np.log(error)
            
            return error
```